File: benchmarks/million_complex.rb
$LOAD_PATH.unshift File.dirname(__FILE__) + '/../lib'
require 'rubygems'
require 'mustache'
 
class MillionComplex < Mustache
  self.path = File.dirname(__FILE__)
 
  def header
    "Colors"
  end
 
  def item
    items = []
    items << { :name => 'red', :current => true, :url => '#Red' }
    items << { :name => 'green', :current => false, :url => '#Green' }
    items << { :name => 'blue', :current => false, :url => '#Blue' }
    items
  end
 
  def link
    not self[:current]
  end
 
  def list
    not item.empty?
  end
 
  def empty
    item.empty?
  end
end
 
if $0 == __FILE__
  puts MillionComplex.to_html
  
  start = Time.now
  1000000.times {
    MillionComplex.to_html
  }
  diff = Time.now - start
  puts "Time taken: #{diff.to_s} seconds"
end

File: benchmarks/million_complex.js
var sys = require('sys')
    fs = require('fs'),
    Path = require('path'),
    mu = require('../lib/mu')
mu.root(__dirname) // __filename

var complex_context = {
  header: function() {
    return "Colors"
  },
  item: [
    {name: "red", current: true, url: "#Red"},
    {name: "green", current: false, url: "#Green"},
    {name: "blue", current: false, url: "#Blue"}
  ],
  link: function() {
    return this["current"] !== true
  },
  list: function() {
    return this.item.length !== 0
  },
  empty: function() {
    return this.item.length === 0
  }
}

var i = 0
console.log("Starting timer...")
console.time('Total time');
(function next() {
  if (i++ < 1000000)
    mu.render(['million_complex.mustache'], complex_context, null, next)
  else
    console.timeEnd('Total time')
})()


  // var i = 0, d = new Date();
  // (function go() {
  //   if (i++ < RUNS) {
  //     mu.render('complex.html', js).once('end', function () { go(); });
  //   }
  // }())
  // process.addListener('exit', function () {
  //   require('util').debug("Time taken: " + ((new Date() - d) / 1000) + "secs");
  // });
